src/game/puppyprint.c

```diff
@@ -878,148 +1182,269 @@ void print_small_text(s32 x, s32 y, const char *str, s32 align, s32 amount, s32
-            print_set_envcolour(r, g, b, 255);
-        } else if (strncmp((str + i), "<SHAKE>", 7) == 0) { // Toggles text that shakes on the spot. Do it again to disable it.
-            shakeToggle ^= 1;
-        } else if (strncmp((str + i), "<WAVE>",  6) == 0) { // Toggles text that waves around. Do it again to disable it.
-            waveToggle  ^= 1;
+            gDPSetEnvColor(gDisplayListHead++, (Color) r, (Color) g, (Color) b, (Color) gCurrEnvCol[3]); // Don't use print_set_envcolour here, also opt to use alpha value from gCurrEnvCol
+        } else {
+            gDPSetEnvColor(gDisplayListHead++, (Color) gCurrEnvCol[0], (Color) gCurrEnvCol[1], (Color) gCurrEnvCol[2], (Color) gCurrEnvCol[3]); // Reset text to envcolor
         }
+    } else if (len == 7 && strncmp((newStr), "<SHAKE>", 7) == 0) { // Toggles text that shakes on the spot. Do it again to disable it.
+        if (!runCMD)
+            return len;
+
+        shakeToggle ^= 1;
+    } else if (len == 6 && strncmp((newStr), "<WAVE>",  6) == 0) { // Toggles text that waves around. Do it again to disable it.
+        if (!runCMD)
+            return len;
+
+        waveToggle  ^= 1;
+    } else {
+        return 0; // Invalid command string; display everything inside to make this clear to the user.
     }
+
     return len;
 }
 
-void get_char_from_byte(u8 letter, s32 *textX, s32 *textY, s32 *spaceX, s32 *offsetY, s32 font) {
+void font_offsets(s8 *offsetY, s32 font, u8 letter) {
+    switch (font) {
+        case FONT_DEFAULT:
+            switch (letter) {
+                // This is for the letters that sit differently on the line. It just moves them down a bit.
+                case 'g': *offsetY = 1 * textSizeTotal; break;
+                case 'q': *offsetY = 1 * textSizeTotal; break;
+                case 'p': *offsetY = 1 * textSizeTotal; break;
+                case 'y': *offsetY = 1 * textSizeTotal; break;
+            }
+        break;
+        case FONT_PLAIN:
+            switch (letter) {
+                // This is for the letters that sit differently on the line. It just moves them down a bit.
+                case 'g': *offsetY = 3 * textSizeTotal; break;
+                case 'q': *offsetY = 3 * textSizeTotal; break;
+                case 'p': *offsetY = 3 * textSizeTotal; break;
+                case 'y': *offsetY = 2 * textSizeTotal; break;
+            }
+        break;
+        case FONT_VANILLA:
+            switch (letter) {
+                // This is for the letters that sit differently on the line. It just moves them down a bit.
+                case 'g': *offsetY = 2 * textSizeTotal; break;
+                case 'q': *offsetY = 2 * textSizeTotal; break;
+                case 'p': *offsetY = 3 * textSizeTotal; break;
+                case 'y': *offsetY = 3 * textSizeTotal; break;
+            }
+        break;
+    }
+}
+
+void get_char_from_byte(u8 letter, s32 *textX, u8 *spaceX, s8 *offsetY, u8 font) {
     *offsetY = 0;
+    u32 let = letter - '!';
```

Comment:
Lol what on earth is going on in this function

Reply:
Exclamation point is the first valid index in puppyprint's charmap, the resultant letter to print is an offset of this.
I "borrowed" (Read: Blatantly stole) this ascii func from Wiseguy's own implementation, which skips over a few characters, so I also sometimes need to read the raw character, too.